django/forms/forms.py

```diff
@@ -282,43 +291,42 @@ def _html_output(self, normal_row, error_row, row_ender, help_text_html, errors_
                 output.append(str_hidden)
         return mark_safe('\n'.join(output))
 
-    def as_table(self):
-        "Return this form rendered as HTML <tr>s -- excluding the <table></table>."
-        return self._html_output(
-            normal_row='<tr%(html_class_attr)s><th>%(label)s</th><td>%(errors)s%(field)s%(help_text)s</td></tr>',
-            error_row='<tr><td colspan="2">%s</td></tr>',
-            row_ender='</td></tr>',
-            help_text_html='<br><span class="helptext">%s</span>',
-            errors_on_separate_row=False,
-        )
-
-    def as_ul(self):
-        "Return this form rendered as HTML <li>s -- excluding the <ul></ul>."
-        return self._html_output(
-            normal_row='<li%(html_class_attr)s>%(errors)s%(label)s %(field)s%(help_text)s</li>',
-            error_row='<li>%s</li>',
-            row_ender='</li>',
-            help_text_html=' <span class="helptext">%s</span>',
-            errors_on_separate_row=False,
-        )
-
-    def as_p(self):
-        "Return this form rendered as HTML <p>s."
-        return self._html_output(
-            normal_row='<p%(html_class_attr)s>%(label)s %(field)s%(help_text)s</p>',
-            error_row='%s',
-            row_ender='</p>',
-            help_text_html=' <span class="helptext">%s</span>',
-            errors_on_separate_row=True,
-        )
+    def get_context(self):
+        fields = []
+        hidden_fields = []
+        top_errors = self.non_field_errors().copy()
+        for name, bf in self._bound_items():
+            bf_errors = self.error_class(bf.errors, renderer=self.renderer)
+            if bf.is_hidden:
+                if bf_errors:
+                    top_errors += [
+                        _('(Hidden field %(name)s) %(error)s') % {'name': name, 'error': str(e)}
+                        for e in bf_errors
+                    ]
+                hidden_fields.append(bf)
+            else:
+                if not isinstance(str(bf_errors), SafeString):
+                    warnings.warn(
+                        'Returning a plain string from ErrorList is deprecated. '
+                        'Please customize via the template system instead.',
+                        RemovedInDjango50Warning,
+                    )
+                # RemovedInDjango50Warning remove mark_safe call
+                fields.append((bf, mark_safe(str(bf_errors))))
```

Comment:
We seem to be calling `str()` twice here. 

Maybe: 

```suggestion
                errors_str = str(bf_errors)
                if not isinstance(errors_str, SafeString):
                    warnings.warn(
                        'Returning a plain string from ErrorList is deprecated. '
                        'Please customize via the template system instead.',
                        RemovedInDjango50Warning,
                    )
                    errors_str = mark_safe(errors_str)
                # RemovedInDjango50Warning: remove check and replace errors_str 
                # with str(bf_errors)
                fields.append((bf, errors_str))
```

Reply:
Not necessarily for now but, will we add any assert/error checking in there when removing this warning? Do we verify (is there a test) that the output **is** a `SafeString`? (Likely already discussed. 😬)

Reply:
I'm not sure, I'll leave this so it can be considered as part of the final reivew. I've also not as yet included Carlton's suggestion.

Reply:
I've now included Carlton's suggestion.